core/trino-main/src/main/java/io/trino/sql/gen/columnar/DictionaryAwareColumnarFilter.java

```diff
@@ -0,0 +1,156 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.sql.gen.columnar;
+
+import io.trino.operator.project.InputChannels;
+import io.trino.spi.Page;
+import io.trino.spi.block.Block;
+import io.trino.spi.block.DictionaryBlock;
+import io.trino.spi.block.RunLengthEncodedBlock;
+
+import static com.google.common.base.Verify.verify;
+import static java.lang.System.arraycopy;
+
+public final class DictionaryAwareColumnarFilter
```

Comment:
This code does not have the logic to disable dictionary processing when it is ineffective like we do in `DictionaryAwarePageFilter`.  Can you add a todo?

Reply:
The problem with that fallback logic is that it causes profile pollution by running both ValueBlock and DictionaryBlock through the same code.
I would like the producer of DictionaryBlock itself to produce dictionary only when it's beneficial and then we don't need fall back logic here. E.g. Parquet writer produces dictionary encoded data only when there is space saving from it and falls back to original encoding when cardinality gets too big. Parquet reader produces dictionary block only when all pages are dictionary encoded and we know that we will read more values than the size of dictionary after applying parquet page filtering.